### ISSUE TITLE
[dynamo + optim] complex, sparse are not on tracing testing path

### ISSUE THREAD

**Issue by jon-chuang** [pytorch/pytorch#112731]
### 🚀 The feature, motivation and pitch

Even though we allow tracing complex and sparse grads for optimizers, we do not test that these actually work

### Solution
Copy some of the tests from test_optim e.g.
https://github.com/pytorch/pytorch/blob/main/test/optim/test_optim.py#L1503

into https://github.com/pytorch/pytorch/blob/main/test/dynamo/test_optimizers.py

### Alternatives

Not tested

### Additional context

NIL

**Comment by janeyx99** [comment]
I agree testing should be prioritized, otherwise we should not say they work with no evidence.

If they're not expected to work (not sure on sparse progress with torch.compile), we should at least smoothly fall back to eager vs. erroring.

**Comment by janeyx99** [comment]
It may be useful to start by enabling PYTORCH_TEST_WITH_DYNAMO across our eager optim tests.
